src/core/string_map.cc

```diff
@@ -65,13 +65,18 @@ StringMap::~StringMap() {
   Clear();
 }
 
-bool StringMap::AddOrUpdate(string_view field, string_view value, uint32_t ttl_sec) {
+bool StringMap::AddOrUpdate(std::string_view field, std::string_view value, uint32_t ttl_sec,
+                            bool keepttl) {
   // 8 additional bytes for a pointer to value.
   auto [newkey, sdsval_tag] = CreateEntry(field, value, time_now(), ttl_sec);
 
   // Replace the whole entry.
-  sds prev_entry = (sds)AddOrReplaceObj(newkey, sdsval_tag & kValTtlBit);
-  if (prev_entry) {
+  if (sds prev_entry = (sds)AddOrReplaceObj(newkey, sdsval_tag & kValTtlBit); prev_entry) {
+    const bool prev_has_ttl =
+        absl::little_endian::Load64(prev_entry + sdslen(prev_entry) + 1) & kValTtlBit;
+    if (keepttl && prev_has_ttl) {
+      SdsUpdateExpireTime(newkey, ObjExpireTime(prev_entry), 8);
+    }
```

Comment:
I see. Ouch. We had a `DCHECK in ObjExpireTime` which we triggered right ? And then, `ObjUpdateExpireTime` used `now()` which changed the TTL. IMO I think the DCHECK in `ObjExpireTime` is a mistake but let's keep it as is for now.

Reply:
The `DCHECK` wasn't triggered earlier as we had no asserts which go through the code path where `KEEPTTL` is passed but object has no expiry, so the object always had a valid expiry, the new assert added at the end of test now does check this code path. 

Earlier what happened during test failure was:

1. add object with `ttl=100`, expiry calculated to 103 (because `now()` == 3)
2. add object again with `keepttl`, which ended up setting expiry to 103 in error.
3. check failed because `100 != FIELDTTL`